### ISSUE TITLE
[BUG] Memory consumption higher when weight layers are clipped after .step()

### ISSUE THREAD

**Issue by jubueche** [microsoft/DeepSpeed#5564]
**Describe the bug**
I have a function that passes the inputs through the model, computes the loss, and finally computes the backward. After the backward, DeepSpeed potentially steps() the optimizer if the gradient accumulated enough.
After the optimizer called step, I want to iterate over my model's linear layers and clip the weights.
When I add the code where I clip the weights, I eventually go out of GPU memory.

This is the code that produces OOM:
```
model.train()
inputs = self._prepare_inputs(inputs)

from deepspeed.runtime.utils import see_memory_usage
see_memory_usage("Before compute loss", force=True)
with self.compute_loss_context_manager():
    loss = self.compute_loss(model, inputs)
see_memory_usage("After compute loss", force=True)

del inputs
torch.cuda.empty_cache()

if self.args.n_gpu > 1:
    loss = loss.mean()  # mean() to average on multi-gpu parallel training

if self.accelerator.deepspeed_engine_wrapped is not None:
    made_step = self.accelerator.deepspeed_engine_wrapped.engine.is_gradient_accumulation_boundary()
else:
    made_step = True

see_memory_usage("Before backward", force=True)
self.accelerator.backward(loss)
see_memory_usage("After backward", force=True)

# this is what we changed. we clip the weights when we updated the parameters
with torch.no_grad():
    if made_step and hasattr(model, "analog_layers"):
        for analog_layer in model.analog_layers():
            analog_layer.clip_weights()

return loss.detach() / self.args.gradient_accumulation_steps
```
When I remove 
```
# this is what we changed. we clip the weights when we updated the parameters
with torch.no_grad():
    if made_step and hasattr(model, "analog_layers"):
        for analog_layer in model.analog_layers():
            analog_layer.clip_weights()
```
it works fine.

For context, the `clip_weights` function looks like this:
```
def clip_weights(self) -> None:
    """Clip the weights."""
    clip_type = self.rpu_config.clip.type
    clip_sigma = self.rpu_config.clip.sigma
    assert clip_sigma > 0, "Clip sigma must be greater than 0"
    sigma_std = clip_sigma * self.weight.std(None, keepdim=True)
    self.weight.data = self.weight.data.clamp(-sigma_std, sigma_std)
```

**To Reproduce**
This is a little tricky to reproduce. I am open to suggestions how to debug this further.

**Expected behavior**
With the clipping of the weights, I don't get an OOM exception.

**ds_report output**
Please run `ds_report` to give us details about your setup.
```
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-devel package with yum
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
fused_adam ............. [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_lion ............... [NO] ....... [OKAY]
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
evoformer_attn ......... [NO] ....... [NO]
 [WARNING]  NVIDIA Inference is only supported on Ampere and newer architectures
fp_quantizer ........... [NO] ....... [NO]
fused_lamb ............. [NO] ....... [OKAY]
fused_lion ............. [NO] ....... [OKAY]
inference_core_ops ..... [NO] ....... [OKAY]
cutlass_ops ............ [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
ragged_device_ops ...... [NO] ....... [OKAY]
ragged_ops ............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.4
 [WARNING]  using untested triton version (3.0.0+989adb9a29), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/gpfs/u/home/ANFM/ANFMbchl/scratch/miniconda3/envs/torch-nightly/lib/python3.10/site-packages/torch']
torch version .................... 2.4.0.dev20240322
deepspeed install path ........... ['/gpfs/u/scratch/ANFM/ANFMbchl/DeepSpeed/deepspeed']
deepspeed info ................... 0.14.3+488a823, 488a823, master
torch cuda version ............... 12.1
torch hip version ................ None
nvcc version ..................... 12.1
deepspeed wheel compiled w. ...... torch 2.4, cuda 12.1
shared memory (/dev/shm) size .... 377.69 GB
```

**System info (please complete the following information):**
 - OS: [e.g. Ubuntu 18.04]
 - GPU count and types 8xV100s

**Launcher context**
I am launching using `accelerate launch ...`

**Comment by jubueche** [comment]
OK the extra memory from the `clamp` was the issue. I replaced it with the in-place version and it works. Not a bug.
